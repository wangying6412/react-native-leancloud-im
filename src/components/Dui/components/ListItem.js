/**
 * 单个列表项
 *
 * @module ListItem
 * @author lihaitang
 * @version v0.0.1
 * @example
 * ```
    <ListItem
        title="列表标题"
        leftIcon="xxxxx"
        onPress = {
            //function
        }
        rightText="xxxx"
        rightTextColor="#000"
        bottomText="xxxxxxxxx"
        badge = {10}
        leftImage = 'xxxxxxxxxxxxxx',
        rightIcon = "none" || ''
        legal
        checked
    />
 * ```
 */

import React, {
    PropTypes
} from 'react';
import {
    Text,
    View,
    Image,
    TouchableHighlight,
} from 'react-native';

import Icon from 'react-native-vector-icons/Ionicons';

import styles from '../css';

class ListItem extends React.Component{

    static get propTypes(){

        const stringOrElement = PropTypes.oneOfType([PropTypes.string,PropTypes.element]);

        return {
            /**
             * ListItem的标题
             */
            title          : stringOrElement,
            /**
             * ListItem 左侧的图标
             */
            leftIcon       : PropTypes.string,
            /**
             * ListItem 左侧的图片，头像之类的
             */
            leftImage      : PropTypes.string,
            /**
             * ListItem 右侧的文字，“查看更多”之类的
             */
            rightText      : stringOrElement,
            /**
             * ListItem 右侧的文字的颜色
             */
            rightTextColor : PropTypes.string,
            /**
             * ListItem 右侧的图标，默认为向右的箭头,如果设为“none”则不显示图标
             */
            rightIcon      : PropTypes.string,
            /**
             * ListItem 底部的文字
             */
            bottomText     : stringOrElement,
            /**
             * ListItem badge就是气泡数字
             */
            badge          : PropTypes.number,
            /**
             * ListItem 大尺寸的列表项
             */
            legal          : PropTypes.bool,
            /**
             * ListItem 点击事件
             */
            onPress        : PropTypes.func,
            /**
             * ListItem 如果设置了该项，右侧图标会变成复选框
             */
            checked        : PropTypes.bool,
        };
    }

    constructor(props){
        super(props);
    }

    render(){

        const title =  this.props.title || this.props.children ;

        return (
            <TouchableHighlight
                onPress = {this.props.onPress && this.props.onPress()}
            >
                <View style={[styles.listItem,this.props.style]}>
                    {
                        this.props.leftIcon &&
                        <View style={[styles.listItem_icon_wrap]}>
                            <Icon name={this.props.leftIcon} style={[styles.listItem_icon,styles['font_gray']]} />
                        </View>
                    }
                    <View style={
                        [styles.listItem_cotent,this.props.last&&styles.noBorder,
                            this.props.legal && styles.listItem_cotent_legal
                        ]
                    }>

                        {
                            this.props.leftImage &&
                            <Image source={{
                                uri : this.props.leftImage
                            }} style={[
                                styles.listItem_leftImage,
                                this.props.legal && styles.listItem_leftImage_legal
                            ]} />
                        }

                        <View style={[styles.flex1]}>
                            <View style={[styles.flex1,styles.row,
                                Boolean(this.props.bottomText) &&
                                styles.listItem_bottomText_before,
                            ]}>
                                <View style={[styles.listItem_title]}>
                                {
                                    typeof title === 'string' ?
                                    <Text style={[styles['font_gray']]}
                                    numberOfLines={1}>
                                    {title}
                                    </Text>
                                    :
                                    title
                                }
                                </View>
                                <Text
                                    style={[styles.listItem_intro,this.props.rightTextColor && {
                                        color : this.props.rightTextColor
                                    }]}
                                    numberOfLines={1}
                                >
                                    {this.props.rightText}
                                </Text>
                            </View>
                            {
                                Boolean(this.props.bottomText) &&
                                <View style={[styles.flex1,styles.listItem_bottomText]}>
                                {
                                    typeof this.props.bottomText === 'string' ?
                                    <Text style={[styles.font_gray,{
                                        minHeight : 20,
                                    }]}
                                        numberOfLines={1}
                                    >{this.props.bottomText}</Text>
                                    :
                                    this.props.bottomText
                                }
                                </View>
                            }
                        </View>
                        {
                            Boolean(this.props.badge) &&
                            <View style={[styles.listItem_badge_wrap]}>
                                <View style={[styles.listItem_badge]}>
                                    <Text style={[styles.font_white]}>{this.props.badge}</Text>
                                </View>
                            </View>
                        }
                        {
                            (this.props.rightIcon !== 'none'
                            &&
                            typeof this.props.checked !== 'boolean')
                            &&
                            <View style={[styles.listItem_icon_wrap]}>
                                <Icon name={ this.props.rightIcon || 'ios-arrow-forward' } style={[styles.listItem_icon,styles.font_gray,
                                ]} />
                            </View>
                        }
                        {
                            (()=>{
                                if(this.props.checked === false){
                                    return (
                                        <View style={[styles.listItem_icon_wrap]}>
                                            <Icon name="ios-checkmark-outline"
                                                style={[styles.listItem_icon_right,styles.font_gray,
                                            ]} />
                                        </View>
                                    );
                                }else if(this.props.checked === true){
                                    return (
                                        <View style={[styles.listItem_icon_wrap]}>
                                            <Icon name="ios-checkmark-circle"
                                            style={[styles.listItem_icon_right,styles.font_red,
                                            ]} />
                                        </View>
                                    );
                                }
                            })()
                        }
                    </View>
                </View>
            </TouchableHighlight>
        );
    }
}

export default ListItem;




