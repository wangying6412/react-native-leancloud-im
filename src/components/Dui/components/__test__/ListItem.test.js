/* global jest describe test expect */

import React from  'react';
import 'react-native';
import renderer from 'react-test-renderer';

import ListItem from '../ListItem';

describe('ListItem',()=>{
    test('ListItem的快照测试',()=>{
        const cb = jest.fn();
        const component = renderer.create(
            <ListItem
                title="列表标题"
                leftIcon="ios-home"
                onPress = {cb}
                rightText="xxxx"
                rightTextColor="#000"
                bottomText="xxxxxxxxx"
                badge = {10}
                leftImage = 'http://xxxxxxxxxxxxxx.com/xxx.jpg'
                rightIcon = "none"
            />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
