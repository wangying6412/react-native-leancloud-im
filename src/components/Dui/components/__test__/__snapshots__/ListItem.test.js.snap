// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem ListItem的快照测试 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      undefined,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFF",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginLeft": 5,
            "minWidth": 22.5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "backgroundColor": "transparent",
                "fontSize": 22.5,
              },
              Object {
                "color": "#808080",
              },
            ],
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "borderBottomWidth": 0.5,
            "borderColor": "#CCC",
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 0,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 0,
            "minHeight": 45,
          },
          undefined,
          undefined,
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "http://xxxxxxxxxxxxxx.com/xxx.jpg",
          }
        }
        style={
          Array [
            Object {
              "alignSelf": "center",
              "backgroundColor": "#000",
              "borderRadius": 17.5,
              "height": 35,
              "marginRight": 10,
              "width": 35,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "flexDirection": "row",
              },
              Object {
                "paddingBottom": 2.5,
                "paddingTop": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#808080",
                  },
                ]
              }
            >
              列表标题
            </Text>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "flex": 1,
                  "textAlign": "right",
                },
                Object {
                  "color": "#000",
                },
              ]
            }
          >
            xxxx
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "paddingBottom": 5,
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#808080",
                },
                Object {
                  "minHeight": 20,
                },
              ]
            }
          >
            xxxxxxxxx
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginLeft": 10,
              "minWidth": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#de6e00",
                "borderRadius": 20,
                "height": 20,
                "justifyContent": "center",
                "paddingBottom": "*",
                "paddingLeft": 0.5,
                "paddingRight": 10,
                "paddingTop": 0,
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#FFF",
                },
              ]
            }
          >
            10
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
